File: circular_list/circular_list_implementation.py
class CircularArray():
    def __init__(self, size):
        self.size = size
        self.queue = [None for i in range(size)] 
        self.front = self.rear = -1
 
    def pushBack(self, data):

        if ((self.rear + 1) % self.size == self.front): 
            print("Error: Lista llena. No se puede agregar el elemento.")
             
        elif (self.front == -1): 
            self.front = 0
            self.rear = 0
            self.queue[self.rear] = data

        else:
            self.rear = (self.rear + 1) % self.size 
            self.queue[self.rear] = data
    
    def pushFront(self, data):
        if ((self.rear + 1) % self.size == self.front): 
            print("Error: Lista llena. No se puede agregar el elemento.")
             
        elif (self.front == -1): 
            self.front = 0
            self.rear = 0
            self.queue[self.rear] = data

        else:
            self.front = (self.front - 1) % self.size
            #self.rear = (self.rear + 1) % self.size 
            self.queue[self.front] = data
    
    def popBack(self):
        if (self.front == -1):
            print ("Error: Lista vacia. No se puede eliminar el elemento.")

        elif (self.front == self.rear): 
            temp=self.queue[self.front]
            self.front = -1
            self.rear = -1
            #return temp
        
        else:
            temp = self.queue[self.rear]
            self.rear = (self.rear - 1) % self.size
            #return temp
    
    def popFront(self):
        if (self.front == -1):
            print ("Error: Lista vacia. No se puede eliminar el elemento.")

        elif (self.front == self.rear): 
            temp=self.queue[self.front]
            self.front = -1
            self.rear = -1
            #return temp
        
        else:
            temp = self.queue[self.front]
            self.front = (self.front + 1) % self.size
            #return temp
 
    def printf(self):
        if self.front == -1:
            print("Elementos en el arreglo: []")

        elif (self.rear >= self.front):
            print("Elementos en el arreglo: [", 
                                              end = "")
            for i in range(self.front, self.rear + 1):
                if i != self.rear and self.front != -1:
                    print(self.queue[i], end = " ")
                else:
                    print(self.queue[i], end = "")
            print ("]")
 
        else:
            print("Elementos en el arreglo: [", 
                                           end = "")
            for i in range(self.front, self.size):
                print(self.queue[i], end = " ")
            for i in range(0, self.rear + 1):
                if i != self.rear and self.front != -1:
                    print(self.queue[i], end = " ")
                else:
                    print(self.queue[i], end = "")
            print ("]")
    
size = int(input())
lst = CircularArray(size)
running = True      
while running:
    try:
        inp = input()
        inp = inp.split()
        if inp[0] == "1":
            lst.printf()
        elif inp[0] == "2":
            lst.pushBack(inp[1])
        elif inp[0] == "3":
            lst.popBack()
        elif inp[0] == "4":
            lst.pushFront(inp[1])
        elif inp[0] == "5":
            lst.popFront()
    except:
        running = False

File: circular_list/circular_list_implementation.cpp
#include <iostream>
using namespace std;

template <typename T>
class CircularList {
    int front = -1;
    int rear = -1;
    int size;
    T* arr;

public:
    CircularList(int n) {
        size = n;
        arr = new T[n];
    }

    ~CircularList() {
        delete[] arr;
    }

    void pushBack(T data) {
        if ((rear + 1) % size == front) {
            cout << "Error: Lista llena. No se puede agregar el elemento.\n";
        } else if (front == -1) {
            front = 0;
            rear = 0;
            arr[rear] = data;
        } else {
            rear = (rear + 1) % size;
            arr[rear] = data;
        }
    }

    void pushFront(T data) {
        if ((rear + 1) % size == front) {
            cout << "Error: Lista llena. No se puede agregar el elemento.\n";
        } else if (front == -1) {
            front = 0;
            rear = 0;
            arr[rear] = data;
        } else {
            front = (front - 1 + size) % size;  // Decremento circular
            arr[front] = data;
        }
    }

    void popBack() {
        if (front == -1) {
            cout << "Error: Lista vacía. No se puede eliminar el elemento.\n";
        } else if (front == rear) {
            front = -1;
            rear = -1;
        } else {
            rear = (rear - 1 + size) % size;  // Decremento circular
        }
    }

    void popFront() {
        if (front == -1) {
            cout << "Error: Lista vacía. No se puede eliminar el elemento.\n";
        } else if (front == rear) {
            front = -1;
            rear = -1;
        } else {
            front = (front + 1) % size;  // Incremento circular
        }
    }

    void print() {
        if (front == -1) {
            cout << "La lista está vacía.\n";
            return;
        }

        cout << "Elementos en el arreglo: [";
        int i = front;
        while (true) {
            cout << arr[i];
            if (i == rear) break;
            cout << " ";
            i = (i + 1) % size;
        }
        cout << "]\n";
    }
};

int main() {
    CircularList<int>* lst = new CircularList<int>(3);
    lst->pushBack(1);
    lst->pushBack(2);
    lst->print();  // [1 2]
    lst->pushBack(3);  // Error: Lista llena
    lst->popFront();
    lst->print();  // [2]
    lst->pushBack(3);
    lst->print();  // [2 3]
    lst->pushFront(0);
    lst->print();  // [0 2 3]
    delete lst;
    return 0;
}
